### PR TITLE
feat(session-replay-browser): fetch privacy configs remotely and join with local

### PR DESCRIPTION
**Pull request by lewgordon-amplitude**
[AMP-98593](https://amplitude.atlassian.net/browse/AMP-98593)

### Summary

Pulls and joins remote privacy config for the session replay browser package. Also adds a custom config endpoint URL for local testing.

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  No.

[AMP-98593]: https://amplitude.atlassian.net/browse/AMP-98593?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by lewgordon-amplitude** [comment]
Waiting on #745

File: packages/session-replay-browser/src/config/joined-config.ts [hunk 1]
```diff
@@ -22,41 +23,88 @@ export class SessionReplayJoinedConfigGenerator {
...
+    // We always want captureEnabled to be true, unless there's an override
+    // in the remote config.
+    config.captureEnabled = true;
+    let remoteConfig: SessionReplayRemoteConfig | undefined;
     try {
+      if (!this.remoteConfigFetch) {
+        return config;
+      }
+
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
```
Comment (kwalker3690) [inline packages/session-replay-browser/src/config/joined-config.ts:39]: These overrides make me think this was typing as any, which maybe means that the remote config updates weren't built for you locally - after pulling v1.x, run `yarn build` in the `packages/analytics-remote-config` folder and see if this typing gets fixed then

File: packages/session-replay-browser/src/config/joined-config.ts [hunk 2]
```diff
@@ -22,41 +23,88 @@ export class SessionReplayJoinedConfigGenerator {
...
+      if (samplingConfig.sample_rate) {
+        config.sampleRate = samplingConfig.sample_rate;
+      }
+    } else {
+      // If config API response was valid (ie 200), but no config returned, assume that
+      // customer has not yet set up config, and use sample rate from SDK options,
+      // allowing for immediate replay capture
+      config.captureEnabled = true;
+      this.localConfig.loggerProvider.debug(
+        'assuming config API response was valid, but no config returned. session replay capture enabled.',
```
Comment (kwalker3690) [inline packages/session-replay-browser/src/config/joined-config.ts:92]: Nit this log ends up being what the customer sees if they have their log level set high enough, so maybe something a little more customer friendly like:
`Remote config successfully fetched, but no values set for project, Session Replay capture enabled`

File: packages/session-replay-browser/src/config/types.ts [hunk 3]
```diff
@@ -1,20 +1,33 @@
...
+  sample_rate: number;
+  capture_enabled: boolean;
+}
+
+/**
+ * NOTE(lew): defining the type like this so that it's easy to convert to
+ * the local config. This won't handle deep copies (probably), so it may
+ * need to be updated in the future.
+ */
+export type PrivacyConfig = {
```
Comment (kwalker3690) [inline packages/session-replay-browser/src/config/types.ts:13]: What about doing
```
export interface SessionReplayPrivacyConfig = {
  blockSelector?: string | string[];
};

export interface PrivacyConfig extends SessionReplayPrivacyConfig
```
(Naming could also definitely be improved here if you have any ideas)

File: packages/session-replay-browser/src/session-replay.ts [hunk 4]
```diff
@@ -49,6 +49,7 @@ export class SessionReplay implements AmplitudeSessionReplay {
     this.identifiers = new SessionIdentifiers({ sessionId: options.sessionId, deviceId: options.deviceId });
     this.joinedConfigGenerator = await createSessionReplayJoinedConfigGenerator(apiKey, options);
     this.config = await this.joinedConfigGenerator.generateJoinedConfig(this.identifiers.sessionId);
+    this.loggerProvider.debug(JSON.stringify(this.config, null, 2));
```
Comment (kwalker3690) [inline packages/session-replay-browser/src/session-replay.ts:52]: I think we can remove this right?

File: packages/session-replay-browser/src/config/types.ts (hunk 3 above)
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/config/types.ts:13]: At this point I think I might just reuse the type as they should be equivalent.

File: packages/session-replay-browser/src/config/joined-config.ts [hunk 5]
```diff
@@ -22,41 +23,83 @@ export class SessionReplayJoinedConfigGenerator {
...
+        return config;
+      }
 
-      if (samplingConfig && Object.keys(samplingConfig).length > 0) {
-        if (Object.prototype.hasOwnProperty.call(samplingConfig, 'capture_enabled')) {
-          config.captureEnabled = samplingConfig.capture_enabled;
-        } else {
-          config.captureEnabled = false;
-        }
+      const samplingConfig = await this.remoteConfigFetch.getRemoteConfig(
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/config/joined-config.ts:39]: @kwalker3690 I made some typing changes here that fixed my issue, but this also provides strong typing in general without type assertions. LMK what you think!

**Review by kwalker3690 — APPROVED** [review]
Love the type changes, nice work

**Pull request merged**